File: GuideObjC/GuideObjC/ViewControllers/UserProfileController/UserProfileController.m
//
//  UserProfileController.m
//  GuideObjC
//
//  Created by LuzanovRoman on 27.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "UserProfileController.h"
#import "UIViewController+Additions.h"
#import "EGFHumanName+Additions.h"
#import "ProgressController.h"
#import "UIColor+Additions.h"
#import "EdgeDownloader.h"
#import "PostController.h"
#import "FollowButton.h"
#import "ProgressCell.h"
#import "Follows.h"

@interface UserProfileController ()
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UILabel *userNameLabel;
@property (weak, nonatomic) IBOutlet FollowButton *followButton;
@property (weak, nonatomic) IBOutlet UIButton *deleteButton;
@property (retain, nonatomic) EdgeDownloader *feed;
@property (retain, nonatomic) NSMutableDictionary * cellHeights;
@property (retain, nonatomic) UIRefreshControl *refreshControl;
@property (retain, nonatomic) NSString *currentUserId;
@end

@implementation UserProfileController

- (void)viewDidLoad {
    [super viewDidLoad];
    _cellHeights = [NSMutableDictionary dictionary];
    
    _refreshControl = [[UIRefreshControl alloc] init];
    _refreshControl.backgroundColor = [UIColor whiteColor];
    _refreshControl.tintColor = [UIColor hexColor:0x5E66B1];
    _tableView.refreshControl = _refreshControl;
    
    [_tableView registerNib:[UINib nibWithNibName:@"ProgressCell" bundle:nil] forCellReuseIdentifier:@"ProgressCell"];
    [_tableView registerNib:[UINib nibWithNibName:@"PostCell" bundle:nil] forCellReuseIdentifier:@"PostCell"];

    if (_profileUser) {
        _userNameLabel.text = [_profileUser.name fullName];
        _followButton.user = _profileUser;

        [self.graph userObjectWithCompletion:^(NSObject * object, NSError * error) {
            if ([object isKindOfClass:[EGFUser class]]) {
                _currentUserId = ((EGFUser *)object).id;
            }
            _feed = [[EdgeDownloader alloc] initWithSource:_profileUser.id edge:@"posts" expand:@[@"creator",@"image"]];
            _feed.tableView = _tableView;
            [_feed getNextPage];
        }];
    }
}

- (IBAction)followUser:(id)sender {
    if (_profileUser) {
        if ([[Follows shared] followStateForUser:_profileUser] == isFollow) {
            [[Follows shared] unfollowUser:_profileUser completion:^{
                [_followButton checkFollowState];
            }];
        }
        else {
            [[Follows shared] followUser:_profileUser completion:^{
                [_followButton checkFollowState];
            }];
        }
        [_followButton checkFollowState];
    }
}

// MARK:- PostCellDelegate
- (NSString *)authorizedUserId {
    return _currentUserId;
}

- (void)deletePost:(EGFPost *)post {
    if (_currentUserId && post.id) {
        [self showConfirmWithTitle:@"Warning" message:@"Really delete?" action:^{
            [ProgressController show];
            [self.graph deleteObjectWithId:post.id forSource:_currentUserId fromEdge:@"posts" completion:^(id object, NSError * error) {
                [ProgressController hide];
            }];
        }];
    }
}

// MARK:- UITableViewDelegate
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    if (self.refreshControl.isRefreshing) {
        [_feed refreshList];
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        EGFPost * post = (EGFPost *)[_feed objectAtIndex:indexPath.row];
        // Check if we already have the value
        NSNumber * height = _cellHeights[post.id];
        
        if (height) {
            return [height floatValue];
        }
        CGFloat newHeight = [PostCell heightForPost:post];
        _cellHeights[post.id] = [NSNumber numberWithFloat:newHeight];
        return newHeight;
    }
    return 44;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        PostController * controller = [self.storyboard instantiateViewControllerWithIdentifier:@"PostController"];
        controller.post = (EGFPost *)[_feed objectAtIndex:indexPath.row];
        [self.navigationController pushViewController:controller animated:true];
    }
}

// MARK:- UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return section == 0 ? _feed.count : 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1) {
        ProgressCell * cell = [tableView dequeueReusableCellWithIdentifier:@"ProgressCell"];
        cell.indicatorIsHidden = tableView.refreshControl.isRefreshing || [_feed isDownloaded];
        return cell;
    }
    PostCell * cell = [tableView dequeueReusableCellWithIdentifier:@"PostCell"];
    cell.delegate = self;
    cell.post = (EGFPost *)[_feed objectAtIndex:indexPath.row];
    return cell;
}
@end

File: GuideSwift/GuideSwift/ViewControllers/UserProfileController/UserProfileController.swift
//
//  UserProfileController.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 27.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class UserProfileController: BaseController, UITableViewDelegate, UITableViewDataSource, PostCellDelegate {
    
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var userNameLabel: UILabel!
    @IBOutlet weak var followButton: FollowButton!

    fileprivate var cellHeights = [String: CGFloat]()
    fileprivate var refreshControl: UIRefreshControl!
    fileprivate var feed: EdgeDownloader<EGFPost>?
    fileprivate var currentUserId: String?
    
    var profileUser: EGFUser?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.register(UINib(nibName: "ProgressCell", bundle: nil), forCellReuseIdentifier: "ProgressCell")
        tableView.register(UINib(nibName: "PostCell", bundle: nil), forCellReuseIdentifier: "PostCell")
        
        refreshControl = UIRefreshControl()
        refreshControl.backgroundColor = UIColor.white
        refreshControl.tintColor = UIColor.hexColor(0x5E66B1)
        tableView.refreshControl = refreshControl

        userNameLabel.text = profileUser?.name?.fullName()
        followButton.user = profileUser
        
        Graph.userObject { (object, error) in
            self.currentUserId = (object as? EGFUser)?.id
            
            guard let profileUserId = self.profileUser?.id else { return }
            self.feed = EdgeDownloader(withSource: profileUserId, edge: "posts", expand: ["creator", "image"])
            self.feed?.tableView = self.tableView
            self.feed?.getNextPage()
        }
    }
    
    @IBAction func followUser(_ sender: AnyObject) {
        guard let user = profileUser else { return }
        
        if Follows.shared.followState(forUser: user) == .isFollow {
            Follows.shared.unfollow(user: user) { self.followButton.checkFollowState() }
        }
        else {
            Follows.shared.follow(user: user) { self.followButton.checkFollowState() }
        }
        followButton.checkFollowState()
    }
    
    // MARK:- PostCellDelegate
    var authorizedUserId: String? {
        get {
            return currentUserId
        }
    }
    
    func delete(post: EGFPost) {
        guard let postId = post.id, let userId = currentUserId else { return }
        
        showConfirm(withTitle: "Warning", message: "Really delete?") {
            ProgressController.show()
            Graph.deleteObject(withId: postId, forSource: userId, fromEdge: "posts") { (_, error) in
                ProgressController.hide()
            }
        }
    }
    
    // MARK:- UITableViewDelegate
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        if let control = refreshControl, control.isRefreshing == true {
            feed?.refreshList()
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.section == 0 {
            if let post = feed?[indexPath.row], let postId = post.id {
                // Check if we already have the value
                if let value = cellHeights[postId] { return value }
                
                let height = PostCell.height(forPost: post)
                cellHeights[postId] = height
                return height
            }
        }
        return 44
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.section != 0 { return }
        
        if let controller = storyboard?.instantiateViewController(withIdentifier: "PostController") as? PostController {
            controller.currentPost = feed?[indexPath.row]
            navigationController?.pushViewController(controller, animated: true)
        }
    }
    
    // MARK:- UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return section == 0 ? (feed?.count ?? 0) : 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 1 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "ProgressCell") as! ProgressCell
            cell.indicatorIsHidden = tableView.refreshControl!.isRefreshing || (feed?.isDownloaded ?? true)
            return cell
        }
        let cell = tableView.dequeueReusableCell(withIdentifier: "PostCell") as! PostCell
        cell.delegate = self
        cell.post = feed![indexPath.row]
        return cell
    }
}
